core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java

```diff
@@ -30,14 +30,28 @@
  * Native implementation {@link StatsContext} serialization.
  */
 final class StatsSerializer {
+  //  * tag_metadata is one byte and is for encoding metadata about the tag. The
+  //    low 2 bits of this byte are used to encode the type of the value bytes.
+  //    The high 6 bits are reserved for future use. The value_bytes type is
+  //    encoded as:
+  //    00 (value 0): string (UTF-8) encoding
+  //    01 (value 1): integer (varint int64 encoding).
+  //    10 (value 2): boolean format.
+  //    11 (value 3): reserved for future use.
+  //    TODO(songya): Currently we only support encoding on string type.
+  private static final int VALUE_TYPE_STRING = 0;
+  private static final int VALUE_TYPE_INTEGER = 1;
+  private static final int VALUE_TYPE_BOOLEAN = 2;
+
   // Serializes a StatsContext by transforming it into a StatsContextProto. The
   // encoded tags are of the form:
-  //   num_tags [key_len key_bytes value_len value_bytes]*
+  //   [tag_metadata key_len key_bytes value_len value_bytes]*
   static void serialize(StatsContextImpl context, OutputStream output) throws IOException {
     ByteBuffer buffer = ByteBuffer.allocate(1024);
```

Comment:
This is not your issue but, since ByteBuffers don't resize automatically, this code could cause problems. Add a TODO() to fix this issue.